### PR TITLE
Assignment client keep a spare

### PR DESCRIPTION
**Pull request by sethalves**
Fixed broken command-line handling which required reversion of previous attempt.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
OS X Yosemite functional test completed with errors
Download recording here:
https://s3.amazonaws.com/hifi-public/pr-builds/recordings/4335.mov

**Comment by hifi-gustavo** [comment]
Build console output for commit: 0c28e220d14c4489f2840208346b09452e5fe555 available at: https://gist.github.com/0ec9d635ad1d8aa8660c<br/><br/>Installers available for download at:<br>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4335.exe<br>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4335.dmg

**Comment by hifi-gustavo** [comment]
OS X Yosemite functional test completed with errors
Download recording here:
https://s3.amazonaws.com/hifi-public/pr-builds/recordings/4335.mov

**Comment by hifi-gustavo** [comment]
Build console output for commit: 5c5ffc07ab4830518735f159d9388c74412dadeb available at: https://gist.github.com/44f9b7571e2a7b0e4fd6<br/><br/>Installers available for download at:<br>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4335.exe<br>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4335.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 0ce49888ff70885e200a50161a5b513713c276f0 available at: https://gist.github.com/02b60130baf0271858e6<br/><br/>Installers available for download at:<br>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4335.exe<br>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4335.dmg

**Comment by hifi-gustavo** [comment]
OS X Yosemite functional test completed without errors
Download recording here:
https://s3.amazonaws.com/hifi-public/pr-builds/recordings/4335.mov

**Comment by hifi-gustavo** [comment]
Build console output for commit: 128a38100ffd179a79e4af09862efcb55ea8d76a available at: https://gist.github.com/99f6facbcc14ea9b5add<br/><br/>Installers available for download at:<br>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4335.exe<br>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4335.dmg

**Comment by hifi-gustavo** [comment]
OS X Yosemite functional test completed with errors
Download recording here:
https://s3.amazonaws.com/hifi-public/pr-builds/recordings/4335.mov

**Comment by hifi-gustavo** [comment]
Build console output for commit: 9f83eabaadf14ce3d9797c94f8ec679ebfbc558f available at: https://gist.github.com/f1a48284e4c100208246<br/><br/>Installers available for download at:<br>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4335.exe<br>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4335.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 0918d4989b342f9380e77dc6cd2e00cbaa0483df available at: https://gist.github.com/287a8e782afac9c33ac4<br/><br/>Installers available for download at:<br>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4335.exe<br>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4335.dmg

**Comment by hifi-gustavo** [comment]
OS X Yosemite functional test completed with errors
Download recording here:
https://s3.amazonaws.com/hifi-public/pr-builds/recordings/4335.mov

**Comment by hifi-gustavo** [comment]
Build console output for commit: 7c7d727b55a878d4a3610dff083be0067502f5cb available at: https://gist.github.com/37064fd5519736db2675<br/><br/>Installers available for download at:<br>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4335.exe<br>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4335.dmg

**Comment by hifi-gustavo** [comment]
OS X Yosemite functional test completed with errors
Download recording here:
https://s3.amazonaws.com/hifi-public/pr-builds/recordings/4335.mov

File: assignment-client/src/AssignmentClientApp.cpp [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+        numForks = minForks;
+    }
+
+    Assignment::Type requestAssignmentType = Assignment::AllTypes;
+    if (argumentVariantMap.contains(ASSIGNMENT_TYPE_OVERRIDE_OPTION)) {
+        requestAssignmentType = (Assignment::Type) argumentVariantMap.value(ASSIGNMENT_TYPE_OVERRIDE_OPTION).toInt();
+    }
+    if (parser.isSet(clientTypeOption)) {
+        if (numForks || minForks || maxForks) {
+            qCritical() << "don't use -t with forking mode.";
```
Comment (birarda) [inline assignment-client/src/AssignmentClientApp.cpp]: fine for now but I'm sure there will be the case where we want a monitor to spin off children all of the same type

File: assignment-client/src/AssignmentClientApp.cpp [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+    quint16 assignmentServerPort = DEFAULT_DOMAIN_SERVER_PORT;
+    if (argumentVariantMap.contains(ASSIGNMENT_WALLET_DESTINATION_ID_OPTION)) {
+        assignmentServerPort = argumentVariantMap.value(CUSTOM_ASSIGNMENT_SERVER_PORT_OPTION).toString().toUInt();
+    }
+    if (parser.isSet(assignmentServerPortOption)) {
+        assignmentServerPort = parser.value(assignmentServerPortOption).toInt();
+    }
+
+
+
```
Comment (birarda) [inline assignment-client/src/AssignmentClientApp.cpp]: extra spaces here

File: assignment-client/src/AssignmentClientApp.cpp [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+    QString assignmentServerHostname;
+    if (argumentVariantMap.contains(ASSIGNMENT_WALLET_DESTINATION_ID_OPTION)) {
+        assignmentServerHostname = argumentVariantMap.value(CUSTOM_ASSIGNMENT_SERVER_HOSTNAME_OPTION).toString();
+    }
+    if (parser.isSet(assignmentServerHostnameOption)) {
+        assignmentServerHostname = parser.value(assignmentServerHostnameOption);
+    }
+
+
```
Comment (birarda) [inline assignment-client/src/AssignmentClientApp.cpp]: extra space here

File: assignment-client/src/AssignmentClientApp.cpp [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+        if (maxForks && maxForks < numForks) {
+            qCritical() << "--max can't be less than -n";
+            parser.showHelp();
+            Q_UNREACHABLE();
+        }
+    }
+
+
+    if (numForks || minForks || maxForks) {
+        // setup a shutdown event listener to handle SIGTERM or WM_CLOSE for us
```
Comment (birarda) [inline assignment-client/src/AssignmentClientApp.cpp]: seems like the wrong comment

File: assignment-client/src/AssignmentClientApp.cpp [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+        AssignmentClientMonitor monitor(numForks, minForks, maxForks, assignmentPool,
+                                        walletUUID, assignmentServerHostname, assignmentServerPort);
+        exec();
+    } else {
+    // setup a shutdown event listener to handle SIGTERM or WM_CLOSE for us
```
Comment (birarda) [inline assignment-client/src/AssignmentClientApp.cpp]: wrong comment, should be indented

File: assignment-client/src/AssignmentClientMonitor.cpp [hunk 6]
```diff
@@ -12,40 +12,51 @@
...
+
+    nodeList->putLocalPortIntoSharedMemory(ASSIGNMENT_CLIENT_MONITOR_LOCAL_PORT_SMEM_KEY, this);
     
     // use QProcess to fork off a process for each of the child assignment clients
-    for (int i = 0; i < numAssignmentClientForks; i++) {
+    for (unsigned int i = 0; i < _numAssignmentClientForks; i++) {
         spawnChildClient();
     }
+
+    connect(&_checkSparesTimer, SIGNAL(timeout()), SLOT(checkSpares()));
```
Comment (birarda) [inline assignment-client/src/AssignmentClientMonitor.cpp]: use the new connect format here
`connect(&_checkSparesTimer, &QTimer::timeout, this, &AssignmentClientMonitor::checkSpares);

File: assignment-client/src/AssignmentClientApp.cpp (hunk 1 above)
Comment (birarda) [inline assignment-client/src/AssignmentClientApp.cpp]: actually we're using this for scripted assignments right now - can you remove this?

File: libraries/networking/src/LimitedNodeList.h [hunk 7]
```diff
@@ -49,6 +50,11 @@ const char STUN_SERVER_HOSTNAME[] = "stun.highfidelity.io";
 const unsigned short STUN_SERVER_PORT = 3478;
 
 const QString DOMAIN_SERVER_LOCAL_PORT_SMEM_KEY = "domain-server.local-port";
+const QString ASSIGNMENT_CLIENT_MONITOR_LOCAL_PORT_SMEM_KEY = "assignment-client-monitor.local-port";
+
+const char DEFAULT_ASSIGNMENT_CLIENT_MONITOR_HOSTNAME[] = "localhost";
+const unsigned short DEFAULT_ASSIGNMENT_CLIENT_MONITOR_PORT = 40104;
+const unsigned short DEFAULT_ASSIGNMENT_CLIENT_MONITOR_DTLS_PORT = 40105;
```
Comment (birarda) [inline libraries/networking/src/LimitedNodeList.h]: we don't need this one

**Comment by birarda** [comment]
Okay - pending those fixes this is good to go. I am out tomorrow so wanted to get it reviewed.

Somebody can merge this in the morning (with the required fixes) - I think that makes more sense so that @murillodigital and @sethalves are definitely around for any fallout.

**Comment by hifi-gustavo** [comment]
Build console output for commit: c80995170af3a813a8f0885c9c6f825348c9e7c7 available at: https://gist.github.com/3867717f2db9101ee7c4<br/><br/>Installers available for download at:<br>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4335.exe<br>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4335.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: b8322556a65a1c3cb2d15eb0de8354191f2d1b17 available at: https://gist.github.com/b823056bce255c0c9112<br/><br/>Installers available for download at:<br>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4335.exe<br>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4335.dmg

**Comment by PhilipRosedale** [comment]
pulling so we can test in office this morning - Stephen said go for it.

**Comment by OmegaHeron** [comment]
Unfortunately - the issue with having assignments-client of count > needed still leads to domain-server printing once per second;
`[DEBUG] [02/27 11:57:00] Received a request for assignment type 7 from 127.0.0.1:52840`
`[DEBUG] [02/27 11:57:00] Unable to fulfill assignment request of type 7 from 127.0.0.1:52840`

This is a waste of I/O time and will create huge log files full of useless cruft making finding valuable information even more difficult.

**Comment by birarda** [comment]
@sethalves there should be a suppress message API you can use to avoid the above, was originally implemented for packet hash mismatch -> I would suggest every minute or couple minutes we output how many of those requests we got from each type/IP/port combo

**Comment by OmegaHeron** [comment]
I did a hackish workaround of launching with assignment-client --min 4 --max 4 to avoid a spare until it's addressed - thought I want the spares when they don't bomb the log.  It adds up to a lot on a machine running 24/7.

**Comment by birarda** [comment]
agree @OmegaHeron we will implement a fix - appreciate the feedback

**Comment by sethalves** [comment]
https://github.com/highfidelity/hifi/pull/4354

I still see these messages, but at a lower rate?

**Pull request merged**